diff --git a/mujoco_viewer/mujoco_viewer.py b/mujoco_viewer/mujoco_viewer.py
index a86a1d6..d1bcc61 100644
--- a/mujoco_viewer/mujoco_viewer.py
+++ b/mujoco_viewer/mujoco_viewer.py
@@ -18,7 +18,8 @@ class MujocoViewer(Callbacks):
             title="mujoco-python-viewer",
             width=None,
             height=None,
-            hide_menus=False):
+            hide_menus=False,
+            diable_key_callbacks=False):
         super().__init__(hide_menus)
 
         self.model = model
@@ -65,12 +66,13 @@ class MujocoViewer(Callbacks):
             glfw.set_mouse_button_callback(
                 self.window, self._mouse_button_callback)
             glfw.set_scroll_callback(self.window, self._scroll_callback)
-            glfw.set_key_callback(self.window, self._key_callback)
+            if not diable_key_callbacks:
+                glfw.set_key_callback(self.window, self._key_callback)
 
         # create options, camera, scene, context
         self.vopt = mujoco.MjvOption()
         self.cam = mujoco.MjvCamera()
-        self.scn = mujoco.MjvScene(self.model, maxgeom=10000)
+        self.scn = mujoco.MjvScene(self.model, maxgeom=100000) # 10000
         self.pert = mujoco.MjvPerturb()
 
         self.ctx = mujoco.MjrContext(
@@ -132,7 +134,7 @@ class MujocoViewer(Callbacks):
 
         # overlay, markers
         self._overlay = {}
-        self._markers = []
+        self._markers_dict = {}
 
     def add_line_to_fig(self, line_name, fig_idx=0):
         assert isinstance(line_name, str), \
@@ -180,7 +182,55 @@ class MujocoViewer(Callbacks):
         fig.linedata[line_idx][1] = line_data
 
     def add_marker(self, **marker_params):
-        self._markers.append(marker_params)
+        m_id = marker_params.get('id', 0)
+        m_key = f'marker_{m_id}'
+        self._markers_dict[m_key] = marker_params
+
+    def draw_or_update_marker(self, ns: str, m_id: int,  **marker_params):
+        if self.scn.ngeom >= self.scn.maxgeom:
+            raise RuntimeError(
+                'Ran out of geoms. maxgeom: %d' %
+                self.scn.maxgeom)
+
+        m_key = f'{ns}_{m_id}'
+        if m_key not in self.marker_geoms:
+            self.marker_geoms[m_key] = self.scn.geoms[self.scn.ngeom]
+            self.scn.ngeom += 1
+
+        m_geom = self.marker_geoms[m_key]
+        m_geom.dataid = -1
+        m_geom.objtype = mujoco.mjtObj.mjOBJ_UNKNOWN
+        m_geom.objid = -1
+        m_geom.category = mujoco.mjtCatBit.mjCAT_DECOR
+        m_geom.emission = 0
+        m_geom.specular = 0.5
+        m_geom.shininess = 0.5
+        m_geom.reflectance = 0
+        m_geom.type = mujoco.mjtGeom.mjGEOM_BOX
+        m_geom.size[:] = np.ones(3) * 0.1
+        m_geom.mat[:] = np.eye(3)
+        m_geom.rgba[:] = np.ones(4)
+
+        for key, value in marker_params.items():
+            if isinstance(value, (int, float, mujoco._enums.mjtGeom)):
+                setattr(m_geom, key, value)
+            elif isinstance(value, (tuple, list, np.ndarray)):
+                attr = getattr(m_geom, key)
+                attr[:] = np.asarray(value).reshape(attr.shape)
+            elif isinstance(value, str):
+                assert key == "label", "Only label is a string in mjtGeom."
+                if value is None:
+                    m_geom.label[0] = 0
+                else:
+                    m_geom.label = value
+            elif hasattr(m_geom, key):
+                raise ValueError(
+                    "mjtGeom has attr {} but type {} is invalid".format(
+                        key, type(value)))
+            else:
+                raise ValueError("mjtGeom doesn't have field %s" % key)
+        return
+
 
     def _add_marker_to_scene(self, marker):
         if self.scn.ngeom >= self.scn.maxgeom:
@@ -194,10 +244,10 @@ class MujocoViewer(Callbacks):
         g.objtype = mujoco.mjtObj.mjOBJ_UNKNOWN
         g.objid = -1
         g.category = mujoco.mjtCatBit.mjCAT_DECOR
-        g.texid = -1
-        g.texuniform = 0
-        g.texrepeat[0] = 1
-        g.texrepeat[1] = 1
+        # g.texid = -1
+        # g.texuniform = 0
+        # g.texrepeat[0] = 1
+        # g.texrepeat[1] = 1
         g.emission = 0
         g.specular = 0.5
         g.shininess = 0.5
@@ -208,6 +258,8 @@ class MujocoViewer(Callbacks):
         g.rgba[:] = np.ones(4)
 
         for key, value in marker.items():
+            if key == "id":
+                continue
             if isinstance(value, (int, float, mujoco._enums.mjtGeom)):
                 setattr(g, key, value)
             elif isinstance(value, (tuple, list, np.ndarray)):
@@ -378,6 +430,7 @@ class MujocoViewer(Callbacks):
             return np.flipud(img)
 
     def render(self):
+        glfw.make_context_current(self.window)
         if self.render_mode == 'offscreen':
             raise NotImplementedError(
                 "Use 'read_pixels()' for 'offscreen' mode.")
@@ -409,7 +462,7 @@ class MujocoViewer(Callbacks):
                     mujoco.mjtCatBit.mjCAT_ALL.value,
                     self.scn)
                 # marker items
-                for marker in self._markers:
+                for _ , marker in self._markers_dict.items():
                     self._add_marker_to_scene(marker)
                 # render
                 mujoco.mjr_render(self.viewport, self.scn, self.ctx)
@@ -467,9 +520,6 @@ class MujocoViewer(Callbacks):
                 update()
                 self._loop_count -= 1
 
-        # clear markers
-        self._markers[:] = []
-
         # apply perturbation (should this come before mj_step?)
         self.apply_perturbations()
 
